File: android/app/src/main/java/ai/thetahealth/mirobody/data/net/ApiEnvelope.kt
package ai.thetahealth.mirobody.data.net

import kotlinx.serialization.Serializable

@Serializable
data class ApiEnvelope<T>(
    val code: Int = 0,
    val message: String? = null,
    val data: T? = null,
)

class ApiException(
    val apiCode: Int,
    message: String?,
) : RuntimeException(message ?: "API error (code=$apiCode)")

fun <T> ApiEnvelope<T>.unwrap(): T {
    if (code != 0) throw ApiException(code, message)
    return data ?: throw ApiException(code, message ?: "empty response data")
}

fun ApiEnvelope<*>.ensureOk() {
    if (code != 0) throw ApiException(code, message)
}

File: ios/Mirobody/Data/Net/ApiEnvelope.swift
import Foundation

/// Wire envelope used by most of the API: `{ code, message, data }`. Mirrors
/// `data/net/ApiEnvelope.kt`.
struct ApiEnvelope<T: Decodable>: Decodable {
    let code: Int
    let message: String?
    let data: T?

    enum CodingKeys: String, CodingKey { case code, message, data }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.code = try c.decodeIfPresent(Int.self, forKey: .code) ?? 0
        self.message = try c.decodeIfPresent(String.self, forKey: .message)
        self.data = try c.decodeIfPresent(T.self, forKey: .data)
    }
}

/// Thrown when an envelope returns a non-zero `code`.
struct ApiError: Error {
    let apiCode: Int
    let message: String?
}

extension ApiEnvelope {
    /// Returns `data` or throws — non-zero code or missing data is an error.
    func unwrap() throws -> T {
        if code != 0 { throw ApiError(apiCode: code, message: message) }
        guard let data else { throw ApiError(apiCode: code, message: message ?? "empty response data") }
        return data
    }

    /// Validates the code without requiring a payload (for delete-style calls).
    func ensureOk() throws {
        if code != 0 { throw ApiError(apiCode: code, message: message) }
    }
}
